### PR TITLE
ENH: add a renderer to save python-profile prof files

### LINKED ISSUE #64: pstats output
One thing that would be very useful is to be able to produce output from pyinstrument in a Python pstats file. That would allow it to, for example, be passed to `gprof2dot` to produce a call-graph tree. It looks like the file format is pretty simple -- some data in a dictionary gets dumped using marshal.dump (see [here](https://github.com/python/cpython/blob/master/Lib/pstats.py#L179)).

### PR DESCRIPTION
**Pull request by mattip**
Closes #64. Now running `python -m pyinstrument -r prof -o <filename> <script>` will create a file compatible with `python -m profile -o <filename> <script>`. It can be loaded into tools like `gprof2dot` and `snakeviz` for visualization.

Of course the `ncall` number is bogus. I used -1. Since we only know total time, the `percall` time is also not available. Ideas for further tests or correcting the statistics are welcome.

### CODE REVIEW

File: pyinstrument/renderers/profrenderer.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        if frame is None:
+            return {}
+
+        key = (frame.file_path or "", frame.line_no or 0, frame.function)
+        val = [-1, -1, frame.time, frame.total_self_time, {}]
+        if frame.parent:
+            p = frame.parent
+            val[-1][(p.file_path or "", p.line_no or 0, p.function)] = val[:-1]
+        if key in stats:
+            return
```
Comment (mattip) [inline pyinstrument/renderers/profrenderer.py:35]: The code in lines 36-38 below is not used. I thought there would be a way to collect better statistics, but none came to mind.

File: pyinstrument/__main__.py [hunk 2]
```diff
@@ -456,15 +461,17 @@ def get_renderer_class(renderer: str) -> type[renderers.Renderer]:
...
         try:
             return object_with_import_path(renderer)
         except (ValueError, ModuleNotFoundError, AttributeError) as err:
             # ValueError means we failed to import this object
             raise OptionsParseError(
                 f"Failed to find renderer with name {renderer!r}.\n"
-                "Options are text, html, json, speedscope or a Python import path to a Renderer\n"
-                "class.\n"
+                "Options are text, html, json, speedscope, prof or a Python\n"
+                "import path to a Rendererclass.\n"
```
Comment (joerick) [inline pyinstrument/__main__.py:474]:
```suggestion
                "import path to a Renderer class.\n"
```

File: pyinstrument/renderers/profrenderer.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    def __init__(self, **kwargs: Any):
+        super().__init__(**kwargs)
+
+    def render_frame(self, frame: Frame | None, stats: dict) -> dict:
+        if frame is None:
+            return {}
+
+        key = (frame.file_path or "", frame.line_no or 0, frame.function)
+        val = [-1, -1, frame.time, frame.total_self_time, {}]
```
Comment (joerick) [inline pyinstrument/renderers/profrenderer.py:30]:
```suggestion
        # format is (call_time, number_calls, total_time, cumulative_time, callers)
        val = [-1, -1, frame.time, frame.total_self_time, {}]
```

File: pyinstrument/renderers/profrenderer.py (hunk 1 above)
Comment (joerick) [reply pyinstrument/renderers/profrenderer.py:35]: would you mind removing them if so?

**Review by joerick — COMMENTED** [review]
Hey Matti! Thanks for this, it looks good. Would it be possible to add just a test that checks that [`pstats.Stats`](https://docs.python.org/3/library/profile.html#module-pstats) doesn't crash when loading the output of a simple script?

File: pyinstrument/__main__.py (hunk 2 above)
Comment (mattip) [reply pyinstrument/__main__.py]: done

File: pyinstrument/renderers/profrenderer.py (hunk 3 above)
Comment (mattip) [reply pyinstrument/renderers/profrenderer.py:31]: done

**Comment by mattip** [comment]
I responded to the review and added a test

**Comment by mattip** [comment]
I think this is ready for another round of review. Perhaps enabling the CI run would give some confidence that the tests are correct?

**Comment by joerick** [comment]
Thanks and apologies for the delay! I'll try to look in the next few days

File: pyinstrument/__main__.py [hunk 4]
```diff
@@ -325,7 +325,12 @@ def dash_m_callback(option: str, opt: str, value: str, parser: optparse.OptionPa
         session = profiler.stop()
 
     if options.outfile:
-        f = codecs.open(options.outfile, "w", "utf-8")
+        if should_close_f_after_writing:
+            f.close()
```
Comment (joerick) [inline pyinstrument/__main__.py:329]: This seems like it's in the wrong place, perhaps an accidental addition? I think these lines should be removed.

```suggestion
```

Comment (joerick) [reply pyinstrument/__main__.py:329]: Actually, there's something weird going on here... the outfile is opened twice. Must be an existing defect. I'll take a look.

**Review by joerick — COMMENTED** [review]
This is complicated by the fact that the prof renderer outputs bytes, not strs. I'll have a think how we can resolve this neatly.

File: pyinstrument/__main__.py (hunk 4 above)
Comment (joerick) [reply pyinstrument/__main__.py:329]: I've refactored this to remove the preexisting issue, I'll push the changes

File: pyinstrument/renderers/profrenderer.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            return {}
+
+        key = (frame.file_path or "", frame.line_no or 0, frame.function)
+        # Format is (call_time, number_calls, total_time, cumulative_time, callers)
+        val = [-1, -1, frame.time, frame.total_self_time, {}]
+        if frame.parent:
+            p = frame.parent
+            val[-1][(p.file_path or "", p.line_no or 0, p.function)] = val[:-1]
+        if key in stats:
+            return
```
Comment (joerick) [inline pyinstrument/renderers/profrenderer.py]: I think this is an issue - we shouldn't discard frames when they've already appeared in stats, we should combine their data. so times should accumulate. I think I have a fix for this, will push shortly.

**Comment by joerick** [comment]
I've pushed some fixes to this, if you get a sec could you check this is still working for you @mattip? There were a couple issues with the prof rendering as well as some internal issues (like total_self_time always returning zero), but I think I got them all.

File: pyinstrument/renderers/profrenderer.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            number_calls = -1
+            total_time = 0
+            cumulative_time = 0
+            callers: dict[FrameKey, CallerValue] = {}
+        else:
+            call_time, number_calls, total_time, cumulative_time, callers = stats[key]
+
+        # update the total time and cumulative time
+        total_time += frame.total_self_time
+        cumulative_time += frame.time
```
Comment (mattip) [inline pyinstrument/renderers/profrenderer.py:53]: Nice, thanks for fixing that.

**Comment by mattip** [comment]
I think this looks good on a toy script I tried. Is there any end-to-end testing you do offline to verify that the renderers all give similar results?

**Comment by joerick** [comment]
Normally I just play with some of the examples scripts in /examples. I checked these with snakeviz. But it would be good to double check the callers dict somehow, maybe there's some way to visualise the call graph? I seem to remember some tool that had blobs that represent functions and lines representing calls...The other thing we could do is assert more deeply in a unit test, but I'm not sure I'm confident enough how the data should look...! I suppose one could compare a toy script with cProfile to verify. Sent from my phoneOn 26 Apr 2023, at 03:38, Matti Picus ***@***.***> wrote:﻿
I think this looks good on a toy script I tried. Is there any end-to-end testing you do offline to verify that the renderers all give similar results?

**Comment by mattip** [comment]
> maybe there's some way to visualise the call graph

Running the code in the new test, and then using `gprof2dot`
```
gprof2dot -f pstats /tmp/test_renderer.prof | dot -Tpng -o /tmp/output.png
```

I get 
output

**Comment by mattip** [comment]
Sorry, that was cProfile. The ProfRenderer from pyinstrument gives me this. Note the `50% -1` in the call to `e` and `sleep`. I wonder what is wrong

output

**Comment by joerick** [comment]
nice, thank you! well the "-1" is because the number of calls is -1, due to statistical profiling we don't know that. but the 50% below the branches is odd, and looks wrong. Perhaps it is getting confused with the numcalls being -1? we could try 1 to see if that helps. Otherwise it's probably something to do with the callers dict

**Comment by mattip** [comment]
Yes, that did fix the per-cent. The rendering is a bit off, it is hard to read 100%. But that is certainly not pyinstrument's problem :). I think this is about as good as we can do?
output

**Comment by joerick** [comment]
yeah that looks good to me. I'd rather the numcalls said -1 rather than 1 or 0 as it's clear that there is no data.

I think i'd like some more detailed unit tests on the things we've discovered during this PR about the pstats file format. mostly, just basic stuff around frames, timings, and asserting that the fix we made above still holds. If you have time, that would be great, otherwise I'll get around to it at some point.

Something else I'd wondering - is `pstats` a better name for this than `prof`?

**Comment by mattip** [comment]
> is pstats a better name for this than prof?

yes, I am horrible at naming things. I will take a look at adding tests. Since the `stats` output is a dictionary, I think it should be straightforward to assert some of what you suggest.

**Comment by mattip** [comment]
Added some tests and renamed the renderer.

**Comment by mattip** [comment]
Rebased off main, and CI is passing.

**Review by joerick — COMMENTED** [review]
Looking over the diff, still a few instances of the renderer referred to as 'prof' and not pstats

**Comment by mattip** [comment]
Good catch. I changed the file extension for this renderer to `*.pstats`.

**Review by joerick — APPROVED** [review]

**Comment by joerick** [comment]
Looks good! Thanks @mattip !

**Comment by mattip** [comment]
Thanks for shepherding this through to the end.

**Comment by mattip** [comment]
Is there a plan for a new release?

**Comment by joerick** [comment]
Sorry for the wait, gonna get #240 in and then cut a new release

**Comment by mattip** [comment]
Cool, thanks

**Pull request merged**